### ISSUE TITLE
Neo4j 2.3.3 bug : unable to distinct on null

### ISSUE THREAD

**Issue by agjini** [neo4j/neo4j#6918]
Neo4j 2.3.3 seems to introduce a bug on distinct() operator on null values. The query works in 2.3.2.

Example :

MATCH (n:Node)
OPTIONAL MATCH (n)-->(n2:Node2)
RETURN n, collect(distinct(n2))

When the n2 doesn't match and returns null column, the query execution crash on server with a NullPointerException : 

``` json
{
  "results": [],
  "errors": [
    {
      "code": "Neo.DatabaseError.Statement.ExecutionFailure",
      "message": null,
      "stackTrace": "java.lang.NullPointerException\n\tat org.neo4j.cypher.internal.compiler.v2_3.pipes.NiceHasherValue.equals(NiceHasher.scala:70)\n\tat scala.collection.mutable.FlatHashTable$class.findElemImpl(FlatHashTable.scala:128)\n\tat scala.collection.mutable.FlatHashTable$class.containsElem(FlatHashTable.scala:121)\n\tat scala.collection.mutable.HashSet.containsElem(HashSet.scala:40)\n\tat scala.collection.mutable.HashSet.contains(HashSet.scala:57)\n\tat org.neo4j.cypher.internal.compiler.v2_3.pipes.aggregation.DistinctFunction.apply(DistinctFunction.scala:38)\n\tat org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(EagerAggregationPipe.scala:84)\n\tat org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(EagerAggregationPipe.scala:84)\n\tat scala.collection.immutable.List.foreach(List.scala:381)\n\tat org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe$$anonfun$internalCreateResults$1.apply(EagerAggregationPipe.scala:84)\n\tat org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe$$anonfun$internalCreateResults$1.apply(EagerAggregationPipe.scala:80)\n\tat scala.collection.Iterator$class.foreach(Iterator.scala:742)\n\tat scala.collection.AbstractIterator.foreach(Iterator.scala:1194)\n\tat org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe.internalCreateResults(EagerAggregationPipe.scala:80)\n\tat org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:125)\n\tat org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:122)\n\tat org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.createResults(DefaultExecutionResultBuilderFactory.scala:93)\n\tat org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.build(DefaultExecutionResultBuilderFactory.scala:63)\n\tat org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:223)\n\tat org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:207)\n\tat org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anon$2.run(ExecutionPlanBuilder.scala:164)\n\tat org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:200)\n\tat org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper$$anonfun$run$1.apply(CompatibilityFor2_3.scala:200)\n\tat org.neo4j.cypher.internal.compatibility.exceptionHandlerFor2_3$.runSafely(CompatibilityFor2_3.scala:116)\n\tat org.neo4j.cypher.internal.compatibility.CompatibilityFor2_3$ExecutionPlanWrapper.run(CompatibilityFor2_3.scala:199)\n\tat org.neo4j.cypher.internal.PreparedPlanExecution.execute(PreparedPlanExecution.scala:27)\n\tat org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:120)\n\tat org.neo4j.cypher.ExecutionEngine.execute(ExecutionEngine.scala:114)\n\tat org.neo4j.cypher.javacompat.internal.ServerExecutionEngine.executeQuery(ServerExecutionEngine.java:62)\n\tat org.neo4j.server.rest.transactional.TransactionHandle.executeStatements(TransactionHandle.java:328)\n\tat org.neo4j.server.rest.transactional.TransactionHandle.commit(TransactionHandle.java:147)\n\tat org.neo4j.server.rest.web.TransactionalService$2.write(TransactionalService.java:211)\n\tat com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:71)\n\tat com.sun.jersey.core.impl.provider.entity.StreamingOutputProvider.writeTo(StreamingOutputProvider.java:57)\n\tat com.sun.jersey.spi.container.ContainerResponse.write(ContainerResponse.java:302)\n\tat com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1510)\n\tat com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1419)\n\tat com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1409)\n\tat com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:409)\n\tat com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:558)\n\tat com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:733)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:790)\n\tat org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)\n\tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)\n\tat org.neo4j.server.rest.web.CollectUserAgentFilter.doFilter(CollectUserAgentFilter.java:69)\n\tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)\n\tat org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)\n\tat org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)\n\tat org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)\n\tat org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)\n\tat org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)\n\tat org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)\n\tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)\n\tat org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)\n\tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)\n\tat org.eclipse.jetty.server.Server.handle(Server.java:497)\n\tat org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)\n\tat org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)\n\tat org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)\n\tat org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)\n\tat java.lang.Thread.run(Thread.java:745)\n"
    }
  ]
}
```

**Comment by jexp** [comment]
That's weird, I can't reproduce it with 2.3.3

```
create (:N);
+-------------------+
| No data returned. |
+-------------------+
Nodes created: 1
Labels added: 1

neo4j-sh (?)$ match (n:N) optional match (n)-[:FOO]->(m) return *;
+--------------------+
| m      | n         |
+--------------------+
| <null> | Node[0]{} |
+--------------------+
1 row
neo4j-sh (?)$ match (n:N) optional match (n)-[:FOO]->(m) return n, collect(distinct m);
+---------------------------------+
| n         | collect(distinct m) |
+---------------------------------+
| Node[0]{} | []                  |
+---------------------------------+
1 row
```

**Comment by jexp** [comment]
Could you also show the `return *` without aggregation?

**Comment by agjini** [comment]
Here is a way to reproduce the problem : 

On an empty database, create the following data:

CREATE (s1:Planet {name:"test"});

CREATE (s2:Planet {name:"test"})-[:MdxPath]->(p:Planet);

MATCH (s:Planet {name:"test"})
CREATE (s)-[:Attribute]->(net:Attribute);

And then here is the statement that produces an NPE:

MATCH (s:Planet {name:"test"})
WITH s
OPTIONAL MATCH (s)-[:MdxPath]->(p:Planet)
WITH [id(p),p.path] as pl, s
MATCH (s)-[:Attribute]->(net:Attribute)
OPTIONAL MATCH (s)<-[:Provided]-(c:Counter)
RETURN collect(distinct(pl)), collect(distinct(net)), collect(distinct(c))

**Comment by jexp** [comment]
Ah thanks, so it is an NPE on comparing null nodes. I.e. your `c`.

```
MATCH (s:Planet {name:"test"})
WITH s
OPTIONAL MATCH (s)-[:MdxPath]->(p:Planet)
WITH [id(p),p.path] as pl, s
MATCH (s)-[:Attribute]->(net:Attribute)
OPTIONAL MATCH (s)<-[:Provided]-(c:Counter) 
return *;

+-------------------------------------------------------------+
| c      | net       | pl              | s                    |
+-------------------------------------------------------------+
| <null> | Node[3]{} | [<null>,<null>] | Node[0]{name:"test"} |
| <null> | Node[4]{} | [2,<null>]      | Node[1]{name:"test"} |
+-------------------------------------------------------------+
```

Minimal reproduction:

```
MATCH (s:Planet {name:"test"})
WITH s
OPTIONAL MATCH (s)-[:MdxPath]->(p:Planet)
WITH [id(p),p.path] as pl, s
MATCH (s)-[:Attribute]->(net:Attribute)
OPTIONAL MATCH (s)<-[:Provided]-(c:Counter)
RETURN collect(distinct c);
java.lang.NullPointerException
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.NiceHasherValue.equals(NiceHasher.scala:70)
    at scala.collection.mutable.FlatHashTable$class.findElemImpl(FlatHashTable.scala:128)
    at scala.collection.mutable.FlatHashTable$class.containsElem(FlatHashTable.scala:121)
    at scala.collection.mutable.HashSet.containsElem(HashSet.scala:40)
    at scala.collection.mutable.HashSet.contains(HashSet.scala:57)
```

This one actually works:

```
MATCH (s:Planet {name:"test"})
> WITH s
> OPTIONAL MATCH (s)-[:MdxPath]->(p:Planet)
> WITH [id(p),p.path] as pl, s
> MATCH (s)-[:Attribute]->(net:Attribute)
> OPTIONAL MATCH (s)<-[:Provided]-(c:Counter)
> RETURN collect(distinct(pl)), collect(distinct(net));
+-------------------------------------------------------+
| collect(distinct(pl))        | collect(distinct(net)) |
+-------------------------------------------------------+
| [[<null>,<null>],[2,<null>]] | [Node[3]{},Node[4]{}]  |
+-------------------------------------------------------+
```

```
neo4j-sh (?)$ export STACKTRACES=true
MATCH (s:Planet {name:"test"})
> WITH s
> OPTIONAL MATCH (s)-[:MdxPath]->(p:Planet)
> WITH [id(p),p.path] as pl, s
> MATCH (s)-[:Attribute]->(net:Attribute)
> OPTIONAL MATCH (s)<-[:Provided]-(c:Counter)
> RETURN collect(distinct(pl)), collect(distinct(net)), collect(distinct(c))

java.lang.NullPointerException
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.NiceHasherValue.equals(NiceHasher.scala:70)
    at scala.collection.mutable.FlatHashTable$class.findElemImpl(FlatHashTable.scala:128)
    at scala.collection.mutable.FlatHashTable$class.containsElem(FlatHashTable.scala:121)
    at scala.collection.mutable.HashSet.containsElem(HashSet.scala:40)
    at scala.collection.mutable.HashSet.contains(HashSet.scala:57)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.aggregation.DistinctFunction.apply(DistinctFunction.scala:38)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(EagerAggregationPipe.scala:84)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe$$anonfun$internalCreateResults$1$$anonfun$apply$1.apply(EagerAggregationPipe.scala:84)
    at scala.collection.immutable.List.foreach(List.scala:381)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe$$anonfun$internalCreateResults$1.apply(EagerAggregationPipe.scala:84)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe$$anonfun$internalCreateResults$1.apply(EagerAggregationPipe.scala:80)
    at scala.collection.Iterator$class.foreach(Iterator.scala:742)
    at scala.collection.AbstractIterator.foreach(Iterator.scala:1194)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.EagerAggregationPipe.internalCreateResults(EagerAggregationPipe.scala:80)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:125)
    at org.neo4j.cypher.internal.compiler.v2_3.pipes.PipeWithSource.createResults(Pipe.scala:122)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.createResults(DefaultExecutionResultBuilderFactory.scala:93)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.DefaultExecutionResultBuilderFactory$ExecutionWorkflowBuilder.build(DefaultExecutionResultBuilderFactory.scala:63)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:223)
    at org.neo4j.cypher.internal.compiler.v2_3.executionplan.ExecutionPlanBuilder$$anonfun$getExecutionPlanFunction$1.apply(ExecutionPlanBuilder.scala:207)
```

**Comment by jexp** [comment]
Thanks again for reporting. 
This should be fixed in 2.3.4 and 3.0-RC1
Please check and close this issue.

https://github.com/neo4j/neo4j/pull/6814

**Comment by Mats-SX** [comment]
@agjini I ran the reproduce steps that you posted, and it did not produce an NPE. I will close this as fixed, by the PR linked by @jexp. Please correct me if I am wrong, and we can re-open this issue.

All the best
Mats
